pkg/actuators/machineset/controller_test.go

```diff
@@ -0,0 +1,359 @@
+			Log:    log.Log,
+			getIbmClient: func(_ string, _ ibmcloudproviderv1.IBMCloudMachineProviderSpec) (ibmclient.Client, error) {
+				return mockIBMClient, nil
+			},
+		}
+
+		// mockIBMClient.EXPECT().InstanceGetProfile(gomock.Any()).Return(true, nil).AnyTimes()
+		// gomock.InOrder(
+		// 	mockIBMClient.EXPECT().InstanceGetProfile(gomock.Any()).Return(false, errors.New("Not Found")),
+		// 	// mockIBMClient.EXPECT().InstanceGetProfile(gomock.Any()).Return(false, errors.New("Not Found")),
+		// 	mockIBMClient.EXPECT().InstanceGetProfile(gomock.Any()).Return(true, nil),
+		// )
+		Expect(err).ToNot(HaveOccurred())
+		Expect(r.SetupWithManager(mgr, controller.Options{})).To(Succeed())
+
+		fakeRecorder = record.NewFakeRecorder(1)
+		r.recorder = fakeRecorder
+
+		c = mgr.GetClient()
+		stopMgr = StartTestManager(mgr)
+
+		namespace = &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{GenerateName: "test-machineset-ns-"}}
+		Expect(c.Create(ctx, namespace)).To(Succeed())
+	})
+
+	AfterEach(func() {
+		Expect(deleteMachineSets(c, namespace.Name)).To(Succeed())
+		// mockCtrl.Finish()
+		stopMgr()
+	})
+
+	type reconcileTestCase = struct {
+		profile             string
+		existingAnnotations map[string]string
+		expectedAnnotations map[string]string
+		expectedEvents      []string
+	}
+
+	DescribeTable("when reconciling MachineSets", func(rtc reconcileTestCase) {
+
+		machineSet, err := newTestMachineSet(namespace.Name, rtc.profile, rtc.existingAnnotations)
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(c.Create(ctx, machineSet)).To(Succeed())
+		Eventually(func() map[string]string {
+			m := &machinev1.MachineSet{}
+			key := client.ObjectKey{Namespace: machineSet.Namespace, Name: machineSet.Name}
+			err := c.Get(ctx, key, m)
+			if err != nil {
+				return nil
+			}
+			annotations := m.GetAnnotations()
+			if annotations != nil {
+				return annotations
+			}
+			// Return an empty map to distinguish between empty annotations and errors
+			return make(map[string]string)
+		}, timeout).Should(Equal(rtc.expectedAnnotations))
+
+		// Check which event types were sent
+		Eventually(fakeRecorder.Events, timeout).Should(HaveLen(len(rtc.expectedEvents)))
+		receivedEvents := []string{}
+		eventMatchers := []gtypes.GomegaMatcher{}
+		for _, eachEvent := range rtc.expectedEvents {
+			receivedEvents = append(receivedEvents, <-fakeRecorder.Events)
+			eventMatchers = append(eventMatchers, ContainSubstring(fmt.Sprintf(" %s ", eachEvent)))
+		}
+		Expect(receivedEvents).To(ConsistOf(eventMatchers))
+	},
+		Entry("with no profile set", reconcileTestCase{
+			profile:             "",
+			existingAnnotations: make(map[string]string),
+			expectedAnnotations: make(map[string]string),
+			expectedEvents:      []string{"FailedUpdate"},
+		}),
+		Entry("with a bx2d-4x16", reconcileTestCase{
+			profile:             "bx2d-4x16",
+			existingAnnotations: make(map[string]string),
+			expectedAnnotations: map[string]string{
+				profileKey: "bx2d-4x16",
```

Comment:
We need the annotations to include the CPU and memory, else the autoscaler doesn't know what to do here. Can this information be retrieved from the profile URL?

Reply:
Yup, udpated.